stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -457,6 +457,27 @@ end
 
 *(A::HermOrSym, B::HermOrSym) = A * copyto!(similar(parent(B)), B)
 
+function dot(x::AbstractVector, A::RealHermSymComplexHerm, y::AbstractVector)
+    require_one_based_indexing(x, y)
+    (length(x) == length(y) == size(A, 1)) || throw(DimensionMismatch())
+    Δ = A.uplo == 'U' ? UpperTriangular(A) : LowerTriangular(A)
+    D = Diagonal(A)
+    if x === y
+        r = 2real(dot(x, Δ, x))
+        @inbounds @simd for i in eachindex(x)
+            r -= dot(x[i], A[i,i], x[i])
```

Comment:
This is not type-stable for a complex-Hermitian matrix, since `A[i,i] <: Complex` even though it is actually real.  Probably you want `real(A.data[i,i])` here (or `real(A[i,i])`, but that is slower because it performs additional checks on the indices.

Reply:
...and then you have to `return complex(r)` so that the return value is type-stable

Reply:
Hm, but in the case of block matrices, `A[i,i]` is `Hermitian(A.data[i,i|)`, where `A.data[i,i]` doesn't have to be `Hermitian` itself...? Then I need to symmetrize somehow?

Reply:
I must be missing something about the `return complex(r)`. If any `eltype` is complex, then we initialize `r` as complex, and then add complex numbers to it, and finally return a complex number. `dot(x[i], A[i,i], x[i])` is not `Real`, and in fact the imaginary part is not exactly 0. I slightly adopted your loop-version below, and tested for inference and it passed. I'll push that and then we may continue based on the update, if necessary.